speedy/model/model.jl

```diff
@@ -0,0 +1,820 @@
+* `dx::AbstractFloat` : Distance between grid points in the lon direction
+* `dy::AbstractFloat` : Distance between grid points in the lat direction
+* `nlev::Int`: Number of points in the vertical direction
+* `n_state_var::Int`: Number of variables in the state vector
+"""
+
+Base.@kwdef struct ModelParameters{T<:AbstractFloat}
+    # Initial date
+    IDate::String=""
+    # Incremental date
+    dtDate::String=""
+    # Hour interval
+    Hinc::Int = 6
+    # Ensemble initialisation
+    ensDate::String=""
+    # Choose observation network (choose "real" or "uniform")
+    obs_network::String = "uniform"
+    # Number of obs stations
+    nobs::Int = 50
+
+    lambda::Vector{T} = [1.0e0, 1.0e0, 1.0e0, 1.0e0, 1.0e0]
+    nu::Vector{T} = [2.5, 2.5, 2.5, 2.5, 2.5]
+    sigma::Vector{T} = [0.1, 0.1, 0.1, 0.00001, 1000.0]
+
+    lambda_initial_state::Vector{T} = [1.0e0, 1.0e0, 1.0e0, 1.0e0, 1.0e0]
+    nu_initial_state::Vector{T} = [2.5, 2.5, 2.5, 2.5, 2.5]
+    sigma_initial_state::Vector{T} = [0.1, 0.1, 0.1, 0.0001, 1000.0]
+    
+    padding::Int = 100
+    primes::Bool = true
+
+    state_prefix::String = "data"
+    title_avg::String = "avg"
+    title_var::String = "var"
+    title_syn::String = "syn"
+    title_grid::String = "grid"
+    title_stations::String = "stations"
+    title_params::String = "params"
+    particle_dump_file = "particle_dump.h5"
+    particle_dump_time = [-1]
+    obs_noise_std::T = 1000.0
+    #Path to the the local speedy directory
+    SPEEDY::String = ""
+    station_filename::String = joinpath(SPEEDY, "obs", "networks", obs_network * ".txt")
+    nature_dir::String = joinpath(SPEEDY, "DATA", "nature")
+    # Output folders
+    output_folder::String = joinpath(pwd(), "speedy")
+    guess_folder::String = joinpath(output_folder, "DATA", "ensemble", "gues")
+    anal_folder::String = joinpath(output_folder, "DATA", "ensemble", "anal")
+    # Assimilated indices
+    assimilate_indices::Vector{Int} = [1,5]
+    # Grid dimensions
+    nlon::Int = 96
+    nlat::Int = 48
+    nlev::Int = 8
+    lon_length::T = 360.0
+    lat_length::T = 180.0
+    z_length::T = 1.0
+    dx::T = lon_length / (nlon - 1)
+    dy::T = lat_length / (nlat - 1)
+    dz::T = z_length / (nlev - 1)
+
+    nij0::Int = nlon*nlat
+    n_state_var::Int = 6
+    n_2d::Int = 2
+    n_3d::Int = 4
+
+end
+function step_datetime(idate::String,dtdate::String)
+    new_idate = Dates.format(DateTime(idate, "YYYYmmddHH") + Dates.Hour(6),"YYYYmmddHH")
+    new_dtdate = Dates.format(DateTime(dtdate, "YYYYmmddHH") + Dates.Hour(6),"YYYYmmddHH")
+    return new_idate,new_dtdate
+end
+
+function step_ens(ensdate::String, Hinc::Int)
+    ens_end = Dates.format(DateTime(ensdate, "YYYYmmddHH") + Dates.Month(1),"YYYYmmddHH")
+    diff = length(DateTime(ensdate, "YYYYmmddHH"):Hour(Hinc):DateTime(ens_end, "YYYYmmddHH"))
+    rand_int = rand(0:diff-1)*Hinc
+    new_idate = Dates.format(DateTime(ensdate, "YYYYmmddHH") + Dates.Hour(rand_int),"YYYYmmddHH")
```

Comment:
If you don't want to repeat `YYYYmmddHH` everywhere, you can do something like
```suggestion

const SPEEDY_DATE_FORMAT = "YYYYmmddHH"

function step_datetime(idate::String,dtdate::String)
    new_idate = Dates.format(DateTime(idate, SPEEDY_DATE_FORMAT) + Dates.Hour(6), SPEEDY_DATE_FORMAT)
    new_dtdate = Dates.format(DateTime(dtdate, SPEEDY_DATE_FORMAT) + Dates.Hour(6), SPEEDY_DATE_FORMAT)
    return new_idate,new_dtdate
end

function step_ens(ensdate::String, Hinc::Int)
    ens_end = Dates.format(DateTime(ensdate, SPEEDY_DATE_FORMAT) + Dates.Month(1), SPEEDY_DATE_FORMAT)
    diff = length(DateTime(ensdate, SPEEDY_DATE_FORMAT):Hour(Hinc):DateTime(ens_end, SPEEDY_DATE_FORMAT))
    rand_int = rand(0:diff-1)*Hinc
    new_idate = Dates.format(DateTime(ensdate, SPEEDY_DATE_FORMAT) + Dates.Hour(rand_int), SPEEDY_DATE_FORMAT)
```